Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+#include "MCHRawElecMap/Mapper.h"
+#include "MCHMappingInterface/Segmentation.h"
+#include "MCHWorkflow/DataDecoder.h"
+
+static bool mPrint = false;
+
+namespace o2
+{
+namespace mch
+{
+namespace raw
+{
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::mch::mapping;
+using RDHv4 = o2::header::RAWDataHeaderV4;
+
+static std::array<int, 64> refManu2ds_st345_v5 = {
+  63, 62, 61, 60, 59, 57, 56, 53, 51, 50, 47, 45, 44, 41, 38, 35,
+  36, 33, 34, 37, 32, 39, 40, 42, 43, 46, 48, 49, 52, 54, 55, 58,
+  7, 8, 5, 2, 6, 1, 3, 0, 4, 9, 10, 15, 17, 18, 22, 25,
+  31, 30, 29, 28, 27, 26, 24, 23, 20, 21, 16, 19, 12, 14, 11, 13};
+
+static std::array<int, 64> refManu2ds_st345_v2 = {
+  62, 61, 63, 60, 59, 55, 58, 57, 56, 54, 50, 46, 42, 39, 37, 41,
+  35, 36, 33, 34, 32, 38, 43, 40, 45, 44, 47, 48, 49, 52, 51, 53,
+  7, 6, 5, 4, 2, 3, 1, 0, 9, 11, 13, 15, 17, 19, 21, 23,
+  31, 30, 29, 28, 27, 26, 25, 24, 22, 20, 18, 16, 14, 12, 10, 8};
+
+#define refManu2ds_st345 refManu2ds_st345_v5
+
+static std::array<int, 64> refDs2manu_st345;
+
+static int manu2ds(int i)
+{
+  return refManu2ds_st345[i];
+}
+
+static int ds2manu(int i)
+{
+  return refDs2manu_st345[i];
+}
+
+class MergerDigit
+{
+ public:
+  MergerDigit() = default;
+  ~MergerDigit() = default;
+
+  o2::mch::Digit digit;
+  o2::mch::Digit::Time stopTime;
+  bool merged = {false};
+  int solarId = {-1};
+  int dsAddr = {-1};
+  int chAddr = {-1};
+};
+
+//_________________________________________________________________
+struct MergerBuffer {
+  std::vector<MergerDigit> digits;
+  uint32_t orbit = {0};
+};
+
+//_________________________________________________________________
+class FeeIdMerger
+{
+  MergerBuffer buffers[2];
+  int currentBufId = {1};
+  int previousBufId = {0};
+  int feeId = {0};
+
+  std::function<void(const Digit&)> sendDigit;
+
+ public:
+  FeeIdMerger() = default;
+  ~FeeIdMerger() = default;
+
+  void setId(int id)
+  {
+    feeId = id;
+  }
+
+  void setDigitHandler(std::function<void(const Digit&)> h)
+  {
+    sendDigit = h;
+  }
+
+  void setOrbit(uint32_t orbit, bool stop);
+
+  MergerBuffer& getCurrentBuffer()
+  {
+    return buffers[currentBufId];
+  }
+
+  MergerBuffer& getPreviousBuffer()
+  {
+    return buffers[previousBufId];
+  }
+
+  int getCurrentBufId()
+  {
+    return currentBufId;
+  }
+
+  int getPreviousBufId()
+  {
+    return previousBufId;
+  }
+
+  void mergeDigits();
+};
+
+//_________________________________________________________________
+class MergerBase
+{
+ public:
+  MergerBase() = default;
+  ~MergerBase() = default;
+
+  virtual void setDigitHandler(std::function<void(const Digit&)> h) = 0;
+  virtual void setOrbit(int feeId, uint32_t orbit, bool stop) = 0;
+  virtual void addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                        int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime) = 0;
+  virtual void mergeDigits(int feeId) = 0;
+};
+
+class MergerSimple : public MergerBase
```

Comment:
if I understand correctly the simple merger is actually not merging at all, right ?

Reply:
Right. We currently only merge data coming from the User Logic, and we should actually only do that when running in continuous mode. In all other cases one can spare some CPU time by simply bypassing the merging algorithm.

Reply:
ok, then it's more than a SimpleMerger, it's a NoOpMerger ;-)

Reply:
I renamed the class NoOpMerger for clarity...